### PR TITLE
Add simple bisimulation prover

### PR DESCRIPTION
**Pull request by bboston7**
This change adds command `prove_bisim` that proves that two terms simulate eachother. Given a relation `rel`, term `lhs`, and term `rhs`, the prover considers `lhs` and `rhs` bisimilar when:
```
forall s1 s2 in out1 out2.
  rel (s1, out1) (s2, out2) -> rel (lhs (s1, in)) (rhs (s2, in))
```

### CODE REVIEW

File: src/SAWScript/Builtins.hs [hunk 1]
```diff
@@ -1226,6 +1227,151 @@ provePrim script t = do
...
+          unless (o1 == o2) $ fail $ unlines
+            [ "Error: Mismatched output types in relation."
+            , "LHS output type: " ++ Cryptol.pretty o1
+            , "RHS output type: " ++ Cryptol.pretty o2 ]
+
+          return (s1, s2, o1)
+        _ -> fail $ "Error: Unexpected relation type: "
+                 ++ show (ppTypedTermType (ttType relation))
+
+    -- Typecheck bisimulation term. Returns the term's output type.
```
Comment (RyanGlScott) [inline src/SAWScript/Builtins.hs:1325]: Similarly, include a comment stating what type you are checking against here.

File: src/SAWScript/Builtins.hs [hunk 2]
```diff
@@ -1226,6 +1227,151 @@ provePrim script t = do
...
+
+    -- Given a 'LocalName' and 'Type' for a function argument, return the same
+    -- 'LocalName' and a 'Term' for the argument type. This function is intended
+    -- to be used in conjunction with 'scLambdaList'.
+    importArg :: SharedContext
+              -> (LocalName, C.Type)
+              -> TopLevel (LocalName, Term)
+    importArg sc (name, t) = do
+      t' <- io $ Cryptol.importType sc Cryptol.emptyEnv t
+      return (name, t')
```
Comment (RyanGlScott) [inline src/SAWScript/Builtins.hs:1373]: I don't think the `name` part of the tuple is used anywhere in the function. Would it be cleaner to remove it and just have the type `SharedContext -> C.Type -> TopLevel Term`?

File: src/SAWScript/Interpreter.hs [hunk 3]
```diff
@@ -1663,6 +1663,23 @@ primitives = Map.fromList
     , "successful, and aborts if unsuccessful."
     ]
 
+  , prim "prove_bisim"         "ProofScript () -> Term -> Term -> Term -> TopLevel ProofResult"
+    (pureVal proveBisimulation)
+    Experimental
+    [ "Use bisimulation to prove that two terms simulate eachother.  The first"
```
Comment (RyanGlScott) [inline src/SAWScript/Interpreter.hs:1669]:
```suggestion
    [ "Use bisimulation to prove that two terms simulate each other.  The first"
```

**Review by RyanGlScott — COMMENTED** [review]
This does what it advertises, so I'm happy with this. I've left a number of suggestions inline, but some of these suggestions may not be applicable, as I'm not completely sure how the bisimulation prover is intended to be used in future work.

As a high-level suggestion: my understanding is that it took a fair bit of discussion and thought to arrive at the current design for the bisimulation prover. I think it would be helpful to capture some of that discussion in a Note somewhere in the comments, including some examples of other design possibilities that were rejected (and with reasons why they were rejected).

File: src/SAWScript/Builtins.hs [hunk 4]
```diff
@@ -1226,6 +1227,151 @@ provePrim script t = do
...
+    fail $ unlines [ "Error: Mismatched input types in bisimulation terms."
+                   , "  LHS input type: " ++ Cryptol.pretty lhsInputType
+                   , "  RHS input type: " ++ Cryptol.pretty rhsInputType ]
+
+  -- Outer function inputs
+  input <- io $ scLocalVar sc 0
+  lhsState <- io $ scLocalVar sc 1
+  rhsState <- io $ scLocalVar sc 2
+  initLhsOutput <- io $ scLocalVar sc 3
+  initRhsOutput <- io $ scLocalVar sc 4
```
[the remark this answers is not in the document]
Comment (bboston7) [reply src/SAWScript/Builtins.hs]: Ah shoot, the Haddocks are wrong. I just fixed them and the help text in `Interpreter.hs` to indicate that there are two initial output values `out1` and `out2`

File: src/SAWScript/Builtins.hs [hunk 5]
```diff
@@ -1226,6 +1227,151 @@ provePrim script t = do
     _ -> return ()
   return res
 
+-- | Use bisimulation to prove that two terms simulate eachother.
+--
+-- Given a relation @rel@, term @lhs@, and term @rhs@, the prover considers
+-- @lhs@ and @rhs@ bisimilar when:
+--   forall s1 s2 in out.
+--     rel (s1, out) (s2, out) -> rel (lhs (s1, in)) (rhs (s2, in))
```
[the remark this answers is not in the document]
Comment (bboston7) [reply src/SAWScript/Builtins.hs]: To your high-level comment, I'll add a detailed NOTE explaining all of this. The short of it is that `out1` and `out2` represent the *initial* output values of the circuit. The idea behind this whole thing is to prove that two circuits executing in lockstep satisfy some relation over the state of each circuit and their outputs. In order to achieve this, we need to prove that if the relation holds at some step of each circuit, and we then step each circuit once, the relation holds after that step. To use the same relation on both sides of this step, the relation consider the output both before and after the step. Thus, we need a universally quantified output from before the step, as well as an input to make the next step. The input and initial outputs are not necessarilly related.

There are other ways to state a different kind of relation that doesn't require an initial output value, but they come with other limitations. I'll spell that all out in the high level NOTE and leave this comment thread unresolved until I've done so.

File: src/SAWScript/Builtins.hs [hunk 6]
```diff
@@ -1226,6 +1227,151 @@ provePrim script t = do
     _ -> return ()
   return res
 
+-- | Use bisimulation to prove that two terms simulate eachother.
+--
+-- Given a relation @rel@, term @lhs@, and term @rhs@, the prover considers
+-- @lhs@ and @rhs@ bisimilar when:
```
[the remark this answers is not in the document]
Comment (bboston7) [reply src/SAWScript/Builtins.hs]: Yeah, I think that would make it clearer. Added.

File: src/SAWScript/Builtins.hs (hunk 2 above)
Comment (bboston7) [reply src/SAWScript/Builtins.hs]: I made this function to try to simplify a `mapM` that needs to preserve the first element in a tuple and modify the second, but I think it wound up being more confusing to follow what's going on. I removed this and just inlined it.

File: src/SAWScript/Builtins.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (bboston7) [reply src/SAWScript/Builtins.hs]: I split `proveBisimulation` into a new module (which we would want anyway as we add more support for bisimulation) and added a module level comment describing the design.

**Review by RyanGlScott — APPROVED** [review]

**Pull request merged**